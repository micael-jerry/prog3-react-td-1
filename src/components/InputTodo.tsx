import React, { useState } from "react";
import { TodoType } from "../types/TodoType";

const InputTodo: React.FC<{
	setTodoList: (todoList: any) => any;
}> = (props) => {
	const { setTodoList } = props;
	const [input, setInput] = useState<TodoType>({
		title: "",
		description: "",
		completed: false,
	});

	const handleSubmit = () => {
		setTodoList((stateTodoList: TodoType[]) => [...stateTodoList, input]);
		setInput({
			title: "",
			description: "",
			completed: false,
		});
	};

	const handleChange = (event: any) => {
		const { value, id } = event.target;
		setInput((inputState: TodoType): TodoType => {
			return {
				...inputState,
				[id]: value,
			};
		});
	};

	return (
		<div id={"inputTodo"}>
			<input
				type={"text"}
				id={"title"}
				onChange={handleChange}
				value={input.title}
			/>
			<input
				type={"text"}
				id={"description"}
				onChange={handleChange}
				value={input.description}
			/>
			<button onClick={handleSubmit}>ADD</button>
		</div>
	);
};

export default InputTodo;
