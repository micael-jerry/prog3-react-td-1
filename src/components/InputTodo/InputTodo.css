.my-btn {
    background-color: #f66969;
    border-color: #f66969;
}

.my-btn:hover {
    background-color: #a33737;
    border-color: #f66969;
}

.form-control:focus {
    box-shadow: 0 0 0 0.25rem #f66969;
}
