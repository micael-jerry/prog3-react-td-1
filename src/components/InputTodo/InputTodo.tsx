import React, { useRef, useState, KeyboardEvent } from "react";
import { TodoType } from "../../types/TodoType";
import "./InputTodo.css";

const InputTodo: React.FC<{
    setTodoList: (todoList: any) => any;
}> = (props) => {
    const { setTodoList } = props;
    const [input, setInput] = useState<TodoType>({
        title: "",
        description: "",
        completed: false,
    });
    const inputTitle = useRef<HTMLInputElement>(null);
    const inputDescription = useRef<HTMLInputElement>(null);

    const handleChange = (event: any) => {
        const { value, id } = event.target;
        setInput((inputState: TodoType): TodoType => {
            return {
                ...inputState,
                [id]: value,
            };
        });
    };

    const handleSubmit = () => {
        setTodoList((stateTodoList: TodoType[]) => [...stateTodoList, input]);
        setInput({
            title: "",
            description: "",
            completed: false,
        });
    };

    const focusDescriptionInputFromTitleInput = (
        event: KeyboardEvent<HTMLInputElement>
    ) => {
        if (event.key === "Enter") {
            inputDescription.current!.focus();
        }
    };

    const focusTitleInputFromDescriptionInput = (
        event: KeyboardEvent<HTMLInputElement>
    ) => {
        if (event.key === "Enter") {
            inputTitle.current!.focus();
        }
    };

    const handleSubmitByPressEnter = (
        event: KeyboardEvent<HTMLInputElement>
    ) => {
        if (event.key === "Enter") {
            handleSubmit();
            focusTitleInputFromDescriptionInput(event);
        }
    };

    return (
        <>
            <div id={"inputTodo"} className={"row align-items-center"}>
                <div className="col-lg-8 col-md-8 col-sm-8">
                    <div className="form-floating mb-4">
                        <input
                            type={"text"}
                            id={"title"}
                            className={"form-control"}
                            ref={inputTitle}
                            value={input.title}
                            onChange={handleChange}
                            onKeyDown={focusDescriptionInputFromTitleInput}
                        />
                        <label htmlFor={"title"}>Title</label>
                    </div>
                    <div className="form-floating mb-4">
                        <input
                            type={"text"}
                            id={"description"}
                            className={"form-control"}
                            ref={inputDescription}
                            value={input.description}
                            onChange={handleChange}
                            onKeyDown={handleSubmitByPressEnter}
                        />
                        <label htmlFor={"description"}>Description</label>
                    </div>
                </div>
                <div className="col-lg-4 col-md-4 col-sm-4">
                    <button
                        onClick={handleSubmit}
                        className={"w-75 btn btn-lg  my-btn mb-3"}
                    >
                        ADD
                    </button>
                </div>
            </div>
        </>
    );
};

export default InputTodo;
