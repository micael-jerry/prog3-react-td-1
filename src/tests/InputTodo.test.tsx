import { fireEvent, render, screen } from "@testing-library/react";
import React from "react";
import InputTodo from "../components/InputTodo/InputTodo";

test("Input title value change", () => {
    const setTodoListMock = jest.fn();
    render(<InputTodo setTodoList={setTodoListMock} />);
    const inputTitle: HTMLInputElement = screen.getByLabelText("title");
    fireEvent.change(inputTitle, {
        target: {
            value: "test title input",
        },
    });

    expect(inputTitle.value).toBe("test title input");
});

test("Input description value change", () => {
    const setTodoListMock = jest.fn();
    render(<InputTodo setTodoList={setTodoListMock} />);
    const inputDescription: HTMLInputElement = screen.getByLabelText("title");
    fireEvent.change(inputDescription, {
        target: {
            value: "test description input",
        },
    });

    expect(inputDescription.value).toBe("test description input");
});

test("Focus on description input from title input after key press Enter", () => {
    const setTodoListMock = jest.fn();
    render(<InputTodo setTodoList={setTodoListMock} />);
    const inputTitle: HTMLInputElement = screen.getByLabelText("title");
    const inputDescription: HTMLInputElement =
        screen.getByLabelText("description");

    fireEvent.keyDown(inputTitle, {
        key: "Enter",
    });
    expect(inputDescription).toHaveFocus();
});

test("Add on Enter key", () => {
    const setTodoListMock = jest.fn();
    render(<InputTodo setTodoList={setTodoListMock} />);
    const inputDescription: HTMLInputElement =
        screen.getByLabelText("description");
    fireEvent.keyDown(inputDescription, {
        key: "Enter",
    });

    expect(setTodoListMock.mock.calls.length).toBe(1);
});

test("Add on ADD click", () => {
    const setTodoListMock = jest.fn();
    render(<InputTodo setTodoList={setTodoListMock} />);
    const addButton = screen.getByText("ADD");
    fireEvent.click(addButton);

    expect(setTodoListMock.mock.calls.length).toBe(1);
});

test("[SNAPSHOT] test render InputTodo", () => {
    const setTodoListMock = jest.fn();
    const { container } = render(<InputTodo setTodoList={setTodoListMock} />);

    expect(container).toMatchSnapshot();
});
