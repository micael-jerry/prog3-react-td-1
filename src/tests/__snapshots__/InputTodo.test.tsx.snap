// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div>
  <div
    class="row align-items-center"
    id="inputTodo"
  >
    <div
      class="col-lg-8 col-md-8 col-sm-8"
    >
      <div
        class="form-floating mb-4"
      >
        <input
          aria-label="title"
          class="form-control"
          id="title"
          type="text"
          value=""
        />
        <label
          for="title"
        >
          Title
        </label>
      </div>
      <div
        class="form-floating mb-4"
      >
        <input
          aria-label="description"
          class="form-control"
          id="description"
          type="text"
          value=""
        />
        <label
          for="description"
        >
          Description
        </label>
      </div>
    </div>
    <div
      class="col-lg-4 col-md-4 col-sm-4"
    >
      <button
        class="w-75 btn btn-lg  my-btn mb-3"
      >
        ADD
      </button>
    </div>
  </div>
</div>
`;

exports[`[SNAPSHOT] test render InputTodo 1`] = `
<div>
  <div
    class="row align-items-center"
    id="inputTodo"
  >
    <div
      class="col-lg-8 col-md-8 col-sm-8"
    >
      <div
        class="form-floating mb-4"
      >
        <input
          aria-label="title"
          class="form-control"
          id="title"
          type="text"
          value=""
        />
        <label
          for="title"
        >
          Title
        </label>
      </div>
      <div
        class="form-floating mb-4"
      >
        <input
          aria-label="description"
          class="form-control"
          id="description"
          type="text"
          value=""
        />
        <label
          for="description"
        >
          Description
        </label>
      </div>
    </div>
    <div
      class="col-lg-4 col-md-4 col-sm-4"
    >
      <button
        class="w-75 btn btn-lg  my-btn mb-3"
      >
        ADD
      </button>
    </div>
  </div>
</div>
`;
